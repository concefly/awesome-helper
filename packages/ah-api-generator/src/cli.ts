#!/usr/bin/env node

import yargs from 'yargs';
import { generateAPIService } from './index';

const args = yargs
  .option('input', { alias: 'i', describe: 'apiDoc', type: 'string', required: true })
  .option('headers', { describe: 'fetch headers', type: 'array' })
  .option('operation-id-splitter', { describe: 'operationId 分隔符，默认', type: 'string' })
  .option('template', { alias: 't', describe: '模板', type: 'string' })
  .option('banner', { alias: 'b', describe: 'banner', type: 'string' })
  .option('dump', { alias: 'd', describe: '导出', type: 'string' }).argv;

generateAPIService({
  input: args.input.match(/^(http|https)\:\/\//)
    ? {
        type: 'remote',
        url: args.input,
        headers: args.headers
          ? args.headers.reduce((re, cu) => {
              const [k, v] = (cu + '').split(':');
              return { ...re, [k]: v };
            }, {} as any)
          : undefined,
      }
    : { type: 'local', filename: args.input },
  operationIdSplitter: args['operation-id-splitter'],
  template: args.template,
  banner: args.banner,
  dump: args.dump,
})
  .then(content => {
    if (!args.dump) console.log(content);
  })
  .catch(err => {
    console.error(err);
    process.exit(1);
  });
