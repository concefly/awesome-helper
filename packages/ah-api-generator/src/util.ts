import { OpenAPIV3 as API } from 'openapi-types';
import _ from 'lodash';
import { Schema } from 'jsonschema';

import { inspect } from 'util';

export const log = (...args: any[]) => {
  console.log(...args.map(arg => inspect(arg, { showHidden: false, depth: null })));
};

export function parseRequestSchema(
  data:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete']
): Schema | undefined {
  if (!data) return;

  const schema: Schema = {
    type: 'object',
    properties: {},
  };

  if (data.parameters) {
    (data.parameters as API.ParameterObject[]).forEach(p => {
      Object.assign(schema.properties!, {
        [p.name]: p.schema,
      });
    });
  }

  if (data.requestBody) {
    const requestBodyContentJson = (data.requestBody as API.RequestBodyObject).content[
      'application/json'
    ];

    if (typeof (requestBodyContentJson?.schema as any)?.type !== 'undefined') {
      const jSchema = requestBodyContentJson?.schema as Schema;
      _.merge(schema, jSchema);
    }
  }

  // 没有收集到，返回空
  if (_.isEmpty(schema.properties)) return;

  return schema;
}

export function parseResponseSchema(
  data:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete']
): Schema | undefined {
  const response = data?.responses?.['200'] as API.ResponseObject;
  const jsonResponse = response.content?.['application/json'];

  if (jsonResponse) {
    // 尝试取 schema
    if (jsonResponse.schema) return jsonResponse.schema as Schema;

    const example =
      jsonResponse.example || (jsonResponse.examples && Object.values(jsonResponse.examples).pop());

    // 尝试转换 example
    if (example) return data2Schema(example);
  }
}

export type IRequestMeta = {
  query?: {
    schema: Schema;
    tsTypeLiteral: string;
  };
  response?: {
    schema: Schema;
    tsTypeLiteral: string;
  };
  pathName: string;
  method: string;
  description?: string;
};

export interface IServiceMap {
  [serviceName: string]: {
    [requestName: string]: IRequestMeta;
  };
}

export function renderService(apiDoc: API.Document, opt: { operationIdSplitter?: string } = {}) {
  const serviceMap: IServiceMap = {};
  const { operationIdSplitter = '.' } = opt;

  Object.entries(apiDoc.paths).forEach(([pathName, pi]) => {
    if (!pi) return;

    Object.entries({
      GET: pi.get,
      PUT: pi.put,
      POST: pi.post,
      DELETE: pi.delete,
    }).forEach(([method, commonPiData]) => {
      if (!commonPiData) return;

      let [serviceName, requestName] = (
        commonPiData.operationId || `default${operationIdSplitter}${_.camelCase(pathName)}`
      ).split(operationIdSplitter);

      serviceName = _.capitalize(serviceName);

      const requestSchema = parseRequestSchema(commonPiData);
      const respSchema = parseResponseSchema(commonPiData);

      const requestMeta: IRequestMeta = {
        query: requestSchema
          ? { schema: requestSchema, tsTypeLiteral: schema2TsTypeLiteral(requestSchema) }
          : undefined,
        response: respSchema
          ? { schema: respSchema, tsTypeLiteral: schema2TsTypeLiteral(respSchema) }
          : undefined,
        method,
        pathName,
        description: commonPiData.description,
      };

      //
      _.set(serviceMap, [serviceName, requestName], requestMeta);
    });
  });

  return serviceMap;
}

export function schema2TsTypeLiteral(s?: Schema): string {
  if (!s) return 'never';

  if (s.type === 'string') return 'string';
  if (s.type === 'integer') return 'number';

  if (s.type === 'array') {
    return `Array<${schema2TsTypeLiteral((s.items as any) || '{}')}>`;
  }

  if (s.type === 'object') {
    {
      if (s.properties) {
        return [
          '{',
          Object.entries(s.properties).map(([pn, pv]) => {
            const isRequired =
              typeof s.required === 'boolean'
                ? s.required
                : Array.isArray(s.required)
                ? s.required.includes(pn)
                : false;

            const tsType = schema2TsTypeLiteral(pv);

            return `${pn}${isRequired ? '' : '?'}: ${tsType}`;
          }),
          '}',
        ].join(' ');
      }

      return 'object';
    }
  }

  return 'any';
}

/** 从数据结构猜测 schema */
export function data2Schema(data: any): Schema {
  if (_.isString(data)) return { type: 'string' };
  if (_.isInteger(data)) return { type: 'integer' };
  if (_.isNumber(data)) return { type: 'number' };

  if (_.isBoolean(data)) return { type: 'boolean' };

  if (_.isArray(data)) {
    return {
      type: 'array',
      items: data2Schema(data[0]),
    };
  }

  if (_.isPlainObject(data)) {
    return {
      type: 'object',
      properties: _.mapValues(data, data2Schema),
      required: Object.keys(data),
    };
  }

  return {};
}
