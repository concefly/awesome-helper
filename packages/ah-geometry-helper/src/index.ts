export * from './IDirType';
export * from './Line';
export * from './LineSegment';
export * from './LineSet';
export * from './Point';
export * from './Rect';
