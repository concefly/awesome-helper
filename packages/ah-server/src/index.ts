import 'koa';
import { DefaultContext, DefaultState, ParameterizedContext } from 'koa';

export interface IApplication {}
export interface IService {}

export interface IContext extends ParameterizedContext<DefaultState, DefaultContext> {}
export interface IConfig {
  LOCAL_PORT: number;
}

// fix koa plugin type
declare module 'koa' {
  interface Request {
    // koaBody type
    body?: any;
  }
}

export * from './App';
export * from './Config';
export * from './Logger';
export * from './Controller';
export * from './Service';
export * from './util';
export * from './Scheduler';
export * from './error';
