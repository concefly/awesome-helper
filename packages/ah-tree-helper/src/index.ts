export * from './TreeHelper';
export * from './Dictionary';
export * from './ITreeNode';
