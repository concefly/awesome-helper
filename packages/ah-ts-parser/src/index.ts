export * from 'ts-morph';
export * from './Ts2Schema';
export * from './util';
