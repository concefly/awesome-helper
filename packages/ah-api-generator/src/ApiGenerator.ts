import _ from 'lodash';
import { OpenAPIV3 as API } from 'openapi-types';
import { IRequestMeta, IServiceMap, renderService } from './util';

interface IConfig {
  renderRequestMethod(ctx: { requestName: string; meta: IRequestMeta }): string;

  renderServiceCls(ctx: {
    StartCaseServiceName: string;
    camelCaseServiceName: string;
    baseServiceName: string;
    requestMethodLiterals: string[];
  }): string;

  renderFileContent(ctx: { serviceClsLiterals: string[]; serviceMap: IServiceMap }): string;
}

/** API 生成器 */
export class ApiGenerator {
  private config: IConfig = {
    renderRequestMethod: ctx => {
      const cDesc = `/** ${ctx.meta.description || ctx.requestName} */`;
      const crCallArgs = `pathName: '${ctx.meta.pathName}', method: '${ctx.meta.method}'`;

      if (!ctx.meta.query) {
        return `
${cDesc}
${ctx.requestName} = async () => { return this.request({ ${crCallArgs} }); }`;
      }

      const schemaLiteral = JSON.stringify(ctx.meta.query.schema);

      return `
${cDesc}
${ctx.requestName} = async (query: ${ctx.meta.query.tsTypeLiteral}) => {
  return this.request({ ${crCallArgs}, query, schema: ${schemaLiteral} });
}`;
    },

    renderServiceCls: ctx => `
export class ${ctx.StartCaseServiceName}Service extends ${ctx.baseServiceName} {
  ${[ctx.requestMethodLiterals.join('\n')]}      
}`,

    renderFileContent: ctx => ctx.serviceClsLiterals.join('\n\n'),
  };

  constructor(readonly apiDoc: API.Document, _config: Partial<IConfig> = {}) {
    Object.assign(this.config, _config);
  }

  private serviceMap = renderService(this.apiDoc);

  /** 导出成字符串 */
  emitToString(opt: { baseServiceName?: string } = {}) {
    const { baseServiceName = 'BaseService' } = opt;

    const serviceClsLiterals = Object.entries(this.serviceMap).map(([serviceName, serviceData]) => {
      const requestMethodLiterals = Object.entries(serviceData).map(([requestName, reqMeta]) => {
        // 生成请求函数
        return this.config.renderRequestMethod({
          requestName,
          meta: reqMeta,
        });
      });

      // 生成 service
      return this.config.renderServiceCls({
        StartCaseServiceName: _.startCase(serviceName).replace(/\s+/g, ''),
        camelCaseServiceName: _.camelCase(serviceName),
        baseServiceName,
        requestMethodLiterals,
      });
    });

    return this.config.renderFileContent({ serviceClsLiterals, serviceMap: this.serviceMap });
  }

  getServiceMap() {
    return _.cloneDeep(this.serviceMap);
  }
}
