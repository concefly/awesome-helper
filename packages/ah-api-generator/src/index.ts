import path from 'path';
import fs from 'fs';
import { curl } from 'urllib';
import { renderService } from './util';
import ejs from 'ejs';
import { OpenAPIV3 as API } from 'openapi-types';
import prettier from 'prettier';
import RefParser from '@apidevtools/json-schema-ref-parser';

export interface IConfig {
  input?: { type: 'local'; filename: string } | { type: 'remote'; url: string; headers?: any };
  operationIdSplitter?: string;
  template?: string;
  banner?: string;
  dump?: string;
}

export const loadApiDoc = async (input: IConfig['input']): Promise<API.Document> => {
  if (!input) throw new Error('input 不存');

  let result =
    input.type === 'local'
      ? JSON.parse(fs.readFileSync(input.filename, { encoding: 'utf-8' }))
      : input.type === 'remote'
      ? await curl<any>(input.url, { headers: input.headers, dataType: 'json' }).then(r => r.data)
      : null;

  if (!result) throw new Error(`inputSource 不存在: ${JSON.stringify(input)}`);

  // 解 $ref
  result = await RefParser.dereference(result);

  return result;
};

export const generateAPIService = async (cfg: IConfig) => {
  const {
    input = { type: 'local', filename: 'api.json' },
    template = path.resolve(__dirname, '..', 'template.ejs'),
    dump,
    banner,
  } = cfg;

  const apiDoc = await loadApiDoc(input);
  const serviceMap = renderService(apiDoc, { operationIdSplitter: cfg.operationIdSplitter });

  // render template
  let content = ejs.render(fs.readFileSync(template, { encoding: 'utf-8' }), { serviceMap });
  if (banner) content = banner + '\n' + content;

  // pretty
  const prettierOpt = await prettier.resolveConfig(process.cwd());
  content = prettier.format(content, {
    parser: 'typescript',
    trailingComma: 'es5',
    tabWidth: 2,
    semi: true,
    singleQuote: true,
    jsxSingleQuote: true,
    printWidth: 100,
    bracketSpacing: true,
    arrowParens: 'avoid',
    ...prettierOpt,
  });

  // dump
  if (dump) fs.writeFileSync(dump, content, { encoding: 'utf-8' });

  return content;
};
