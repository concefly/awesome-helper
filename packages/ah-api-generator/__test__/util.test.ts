import { data2Schema, renderService } from '../src';
import apiJson from './api.json';

describe('util', () => {
  it('renderService', () => {
    const serviceMap = renderService(apiJson as any);
    expect(serviceMap).toMatchSnapshot();
  });

  describe('data2Schema', () => {
    it('normal', () => {
      const s = data2Schema({
        pageSize: 1,
        pageNum: 1,
        total: 783562,
        list: [
          {
            id: 1,
            created_at: '2021-01-24T17:22:45+08:00',
            updated_at: '2021-01-24T17:22:45+08:00',
            level: 0,
            a: '0dfew2',
            b: false,
            c: null,
            d: undefined,
            e: true,
            f: NaN,
          },
        ],
      });

      expect(s).toMatchSnapshot();
    });
  });
});
