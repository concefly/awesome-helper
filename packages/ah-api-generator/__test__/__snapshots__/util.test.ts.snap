// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util renderService 1`] = `
Object {
  "page": Object {
    "createOne": Object {
      "body": "return this.request({ pathName: '/page', method: 'post', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ title: string,summary: string,content: string,meta: object }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "content": Object {
            "type": "string",
          },
          "meta": Object {
            "type": "object",
          },
          "summary": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "required": Array [
          "title",
          "summary",
          "content",
          "meta",
        ],
        "type": "object",
      },
    },
    "getList": Object {
      "body": "return this.request({ pathName: '/page', method: 'get', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ name?: string,title?: string,summary?: string }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "name": Object {
            "type": "string",
          },
          "summary": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    "removeOne": Object {
      "body": "return this.request({ pathName: '/page/:name', method: 'delete', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ name?: string }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "name": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    "updateOne": Object {
      "body": "return this.request({ pathName: '/page/:name', method: 'put', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ name: string,title?: string,summary?: string,content?: string,meta?: object }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "content": Object {
            "type": "string",
          },
          "meta": Object {
            "type": "object",
          },
          "name": Object {
            "type": "string",
          },
          "summary": Object {
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
        },
        "required": Array [
          "name",
        ],
        "type": "object",
      },
    },
  },
  "reserve": Object {
    "confirmUserRequest": Object {
      "body": "return this.request({ pathName: '/reserve_request/:id/confirm', method: 'put', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ id: number,username: string }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "id": Object {
            "type": "integer",
          },
          "username": Object {
            "type": "string",
          },
        },
        "required": Array [
          "id",
          "username",
        ],
        "type": "object",
      },
    },
    "createActivity": Object {
      "body": "return this.request({ pathName: '/reserve_activity', method: 'post', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ title: string,detail?: string,meta?: { richDetailMeta?: object } }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "detail": Object {
            "type": "string",
          },
          "meta": Object {
            "properties": Object {
              "richDetailMeta": Object {
                "type": "object",
              },
            },
            "type": "object",
          },
          "title": Object {
            "description": "活动标题",
            "type": "string",
          },
        },
        "required": Array [
          "title",
        ],
        "type": "object",
      },
    },
    "createCalender": Object {
      "body": "return this.request({ pathName: '/reserve_calendar', method: 'post', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ activity_id: number,list: any }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "activity_id": Object {
            "type": "integer",
          },
          "list": Object {
            "items": Object {
              "properties": Object {
                "detail": Object {
                  "type": "string",
                },
                "end_time": Object {
                  "type": "string",
                },
                "location": Object {
                  "type": "string",
                },
                "meta": Object {
                  "properties": Object {
                    "richDetailMeta": Object {
                      "type": "object",
                    },
                  },
                  "type": "object",
                },
                "start_time": Object {
                  "type": "string",
                },
                "status": Object {
                  "enum": Array [
                    "enabled",
                    "disabled",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "start_time",
                "end_time",
              ],
              "type": "object",
            },
            "maxItems": 30,
            "type": "array",
          },
        },
        "required": Array [
          "activity_id",
          "list",
        ],
        "type": "object",
      },
    },
    "getActivityList": Object {
      "body": "return this.request({ pathName: '/reserve_activity', method: 'get', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ id?: number }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "id": Object {
            "type": "integer",
          },
        },
        "type": "object",
      },
    },
    "getCalenderList": Object {
      "body": "return this.request({ pathName: '/reserve_calendar', method: 'get', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ id?: number,activity_id?: number,startTime?: string,endTime?: string }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "activity_id": Object {
            "type": "integer",
          },
          "endTime": Object {
            "type": "string",
          },
          "id": Object {
            "type": "integer",
          },
          "startTime": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    "getUserRequestList": Object {
      "body": "return this.request({ pathName: '/reserve_request', method: 'get', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ calendar_id?: number }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "calendar_id": Object {
            "type": "integer",
          },
        },
        "type": "object",
      },
    },
    "removeActivity": Object {
      "body": "return this.request({ pathName: '/reserve_activity/:id', method: 'delete', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ id?: number }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "id": Object {
            "type": "integer",
          },
        },
        "type": "object",
      },
    },
    "updateActivity": Object {
      "body": "return this.request({ pathName: '/reserve_activity/:id', method: 'put', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ id: number,title?: string,detail?: string,meta?: { richDetailMeta?: object } }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "detail": Object {
            "type": "string",
          },
          "id": Object {
            "type": "integer",
          },
          "meta": Object {
            "properties": Object {
              "richDetailMeta": Object {
                "type": "object",
              },
            },
            "type": "object",
          },
          "title": Object {
            "type": "string",
          },
        },
        "required": Array [
          "id",
        ],
        "type": "object",
      },
    },
    "updateCalender": Object {
      "body": "return this.request({ pathName: '/reserve_calendar/:id', method: 'put', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ id: number,start_time?: string,end_time?: string,location?: string,detail?: string,status?: string,meta?: { richDetailMeta?: object } }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "detail": Object {
            "type": "string",
          },
          "end_time": Object {
            "type": "string",
          },
          "id": Object {
            "type": "integer",
          },
          "location": Object {
            "type": "string",
          },
          "meta": Object {
            "properties": Object {
              "richDetailMeta": Object {
                "type": "object",
              },
            },
            "type": "object",
          },
          "start_time": Object {
            "type": "string",
          },
          "status": Object {
            "enum": Array [
              "enabled",
              "disabled",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "id",
        ],
        "type": "object",
      },
    },
  },
  "system": Object {
    "getApiDoc": Object {
      "body": "return this.request({ pathName: '/system/apiDoc', method: 'get',  });",
      "parameters": Array [],
      "reqSchema": undefined,
    },
    "getList": Object {
      "body": "return this.request({ pathName: '/system', method: 'get',  });",
      "parameters": Array [],
      "reqSchema": undefined,
    },
    "updateOne": Object {
      "body": "return this.request({ pathName: '/system/:name', method: 'put',  });",
      "parameters": Array [],
      "reqSchema": undefined,
    },
  },
  "user": Object {
    "addReserveRequest": Object {
      "body": "return this.request({ pathName: '/user/reserve_request', method: 'post', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ calendar_id: number,type: string,meta?: { members?: any } }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "calendar_id": Object {
            "type": "integer",
          },
          "meta": Object {
            "properties": Object {
              "members": Object {
                "items": Object {
                  "properties": Object {
                    "gender": Object {
                      "enum": Array [
                        "男",
                        "女",
                      ],
                      "type": "string",
                    },
                    "nick": Object {
                      "type": "string",
                    },
                    "phone": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "nick",
                    "gender",
                    "phone",
                  ],
                  "type": "object",
                },
                "maxItems": 20,
                "type": "array",
              },
            },
            "type": "object",
          },
          "type": Object {
            "enum": Array [
              "team",
              "personal",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "calendar_id",
          "type",
        ],
        "type": "object",
      },
    },
    "assignShiluRoles": Object {
      "body": "return this.request({ pathName: '/user/shiluRoles', method: 'post', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ roleId: number }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "roleId": Object {
            "type": "integer",
          },
        },
        "required": Array [
          "roleId",
        ],
        "type": "object",
      },
    },
    "cancelReserveRequest": Object {
      "body": "return this.request({ pathName: '/user/reserve_request/:id/cancel', method: 'put', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ id: number }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "id": Object {
            "type": "integer",
          },
        },
        "required": Array [
          "id",
        ],
        "type": "object",
      },
    },
    "getEnterCode": Object {
      "body": "return this.request({ pathName: '/user/enterCode', method: 'get', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ pavilionId?: number }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "pavilionId": Object {
            "type": "integer",
          },
        },
        "type": "object",
      },
    },
    "getLeaveInfo": Object {
      "body": "return this.request({ pathName: '/user/leaveInfo', method: 'get', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ pavilionId?: number }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "pavilionId": Object {
            "type": "integer",
          },
        },
        "type": "object",
      },
    },
    "getShiluRoles": Object {
      "body": "return this.request({ pathName: '/user/shiluRoles', method: 'get',  });",
      "parameters": Array [],
      "reqSchema": undefined,
    },
    "getUserInfo": Object {
      "body": "return this.request({ pathName: '/user', method: 'get',  });",
      "parameters": Array [],
      "reqSchema": undefined,
    },
    "removeShiluRoles": Object {
      "body": "return this.request({ pathName: '/user/shiluRoles', method: 'delete', query });",
      "parameters": Array [
        Object {
          "name": "query",
          "tsType": "{ roleId?: number }",
        },
      ],
      "reqSchema": Object {
        "properties": Object {
          "roleId": Object {
            "type": "integer",
          },
        },
        "type": "object",
      },
    },
    "updateSession": Object {
      "body": "return this.request({ pathName: '/user/updateSession', method: 'get',  });",
      "parameters": Array [],
      "reqSchema": undefined,
    },
  },
}
`;
